### PR TITLE
#132 CcXOR is not strong enough

### PR DESCRIPTION
**Pull request by super132**
@davvd , implementation of AES algorithm using JCE has been done for this task #132 . Would you please review? Thanks.

### CODE REVIEW

**Comment by davvd** [comment]
@super132 Thanks, I will find someone to review this PR soon

**Comment by davvd** [comment]
@ggajos please review,thanks

File: src/main/java/org/takes/facets/auth/codecs/CcAES.java [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+
+    /**
+     * The algorithm used.
+     */
+    private static final transient String ALGORITHM = "AES/CBC/PKCS5PADDING";
+
+    /**
+     * Default block size (128 bits) in bytes.
+     */
+    private static final transient int BLOCK = 16;
```
Comment (ggajos) [inline src/main/java/org/takes/facets/auth/codecs/CcAES.java]: @super132 this value is used only in single method, no need to extract this as field.

File: src/main/java/org/takes/facets/auth/codecs/CcAES.java [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+     *
+     * @param bytes Bytes to encrypt
+     * @return Encrypted byte using AES algorithm
+     * @throws IOException for all unexpected exceptions
+     */
+    private byte[] encrypt(final byte[] bytes) throws IOException {
+        try {
+            final Cipher cipher = Cipher.getInstance(ALGORITHM);
+            final AlgorithmParameterSpec spec =
+                    new IvParameterSpec(this.ivbytes);
```
Comment (ggajos) [inline src/main/java/org/takes/facets/auth/codecs/CcAES.java]: @super132 we can init decrypt/encrypt ciphers in Ctor, no need to create this each time.

File: src/main/java/org/takes/facets/auth/codecs/CcAES.java [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+     *
+     * @param bytes Bytes to decrypt
+     * @return Decrypted bytes
+     * @throws IOException for all unexpected exceptions
+     */
+    private byte[] decrypt(final byte[] bytes) throws IOException {
+        try {
+            final Cipher cipher = Cipher.getInstance(ALGORITHM);
+            final AlgorithmParameterSpec spec =
+                    new IvParameterSpec(this.ivbytes);
```
Comment (ggajos) [inline src/main/java/org/takes/facets/auth/codecs/CcAES.java]: @super132 we can init decrypt/encrypt ciphers in Ctor, no need to create this each time.

File: src/main/java/org/takes/facets/auth/codecs/CcAES.java [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+            cipher.init(Cipher.DECRYPT_MODE, this.secret, spec);
+            return cipher.doFinal(bytes);
+        } catch (final InvalidKeyException ike) {
+            throw new IOException(ike);
```
Comment (ggajos) [inline src/main/java/org/takes/facets/auth/codecs/CcAES.java]: @super132 we should not hide all exceptions, why just not throw them?

File: src/main/java/org/takes/facets/auth/codecs/CcAES.java [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+import javax.crypto.SecretKey;
+import javax.crypto.spec.IvParameterSpec;
+import javax.crypto.spec.SecretKeySpec;
+import lombok.EqualsAndHashCode;
+import org.takes.facets.auth.Identity;
+
+/**
+ * AES codec which support 128 bits key.
+ *
+ * <p>The class is immutable and thread-safe.
```
Comment (ggajos) [inline src/main/java/org/takes/facets/auth/codecs/CcAES.java]: @super132 if is immutable we might use `@Immutable` annotation

File: src/test/java/org/takes/facets/auth/codecs/CcAESTest.java [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+     * Default key size.
+     */
+    private static final transient int SIZE = 128;
+
+    /**
+     * Testing AES encryption and decryption.
+     * @throws Exception any unexpected exception to throw
+     */
+    @Test
+    public void encryptAndDecrypt() throws Exception {
```
Comment (ggajos) [inline src/test/java/org/takes/facets/auth/codecs/CcAESTest.java]: @super132 incorrect test name and description http://www.yegor256.com/2014/04/27/typical-mistakes-in-java-code.html#test-method-names

File: src/test/java/org/takes/facets/auth/codecs/CcAESTest.java [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+ * @author Jason Wong (super132j@yahoo.com)
+ * @version $Id$
+ * @since 0.13.8
+ */
+public final class CcAESTest {
+
+    /**
+     * Default key size.
+     */
+    private static final transient int SIZE = 128;
```
Comment (ggajos) [inline src/test/java/org/takes/facets/auth/codecs/CcAESTest.java]: @super132 this variable is used only once, we can inline

File: src/main/java/org/takes/facets/auth/codecs/CcAES.java (hunk 4 above)
Comment (super132) [inline src/main/java/org/takes/facets/auth/codecs/CcAES.java]: @ggajos, this method in turns called by ```encode()``` and ```decode()``` method and it is expected to throw ```IOException```. Therefore, I try to wrap it nicely here to make sure it throws only IOException.

File: src/main/java/org/takes/facets/auth/codecs/CcAES.java (hunk 1 above)
Comment (super132) [inline src/main/java/org/takes/facets/auth/codecs/CcAES.java]: @ggajos, do you mean to declare it as a final variable in the method? As I got complaint from qulice saying that this is a magic number thus I tried extract it out. I will modify it to be inline.

File: src/main/java/org/takes/facets/auth/codecs/CcAES.java (hunk 4 above)
Comment (ggajos) [inline src/main/java/org/takes/facets/auth/codecs/CcAES.java]: @super132 I see, think when we move some code to Ctor this list should shrink a bit

File: src/main/java/org/takes/facets/auth/codecs/CcAES.java (hunk 1 above)
Comment (ggajos) [inline src/main/java/org/takes/facets/auth/codecs/CcAES.java]: @super132 I see, let's inline then and suppress that warning maybe?

File: src/main/java/org/takes/facets/auth/codecs/CcAES.java (hunk 4 above)
Comment (super132) [inline src/main/java/org/takes/facets/auth/codecs/CcAES.java]: @ggajos, right. I will further to catch only ```GeneralSecurityException``` to further shrink the list. Thanks.

**Comment by super132** [comment]
@ggajos, all comments have been addressed. Would you please review? Thanks.

File: src/main/java/org/takes/facets/auth/codecs/CcAES.java [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+        final int block = 16;
+        if (key.length != block) {
+            throw new IllegalArgumentException(
+                "Key length must be in 128-bit."
+            );
+        }
+        this.origin = codec;
+        final byte[] passcode = new byte[key.length];
+        System.arraycopy(key, 0, passcode, 0, key.length);
+        final SecureRandom random = new SecureRandom();
```
Comment (ggajos) [inline src/main/java/org/takes/facets/auth/codecs/CcAES.java]: @super132 let's not declare variables that are used only one time, let's inline it

**Comment by ggajos** [comment]
@super132 just one more comment and we good to go

File: src/main/java/org/takes/facets/auth/codecs/CcAES.java (hunk 8 above)
Comment (super132) [inline src/main/java/org/takes/facets/auth/codecs/CcAES.java]: @ggajos, may I know how to suppress qulice complains if inline a number more than once? Thanks.

Comment (ggajos) [inline src/main/java/org/takes/facets/auth/codecs/CcAES.java]: @super132 not sure If I got it right but current (commited) approach is correct

**Comment by ggajos** [comment]
@super132 looks like build is failing due to:
```
[WARNING] Used undeclared dependencies found:
	com.jcabi:jcabi-aspects:jar:0.20.1:compile
```

**Comment by super132** [comment]
@ggajos, I think the failure is due to a known bug. Let me try recommiting again and let the build run again.

**Comment by super132** [comment]
@ggajos, same failure. May I know how can we fix the undeclared dependency problem?

**Comment by super132** [comment]
@ggajos , build is fine with updated POM, please review. Thanks.

**Comment by ggajos** [comment]
@super132 thanks, looks good now

**Comment by ggajos** [comment]
@rultor merge please

File: src/main/java/org/takes/facets/auth/codecs/CcAES.java [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+     * Ctor.
+     * @param codec Original codec
+     * @param key The encryption key
+     * @exception IOException errors on creating internal components
+     */
+    public CcAES(final Codec codec, final byte[] key) throws IOException {
+        final int block = 16;
+        if (key.length != block) {
+            throw new IllegalArgumentException(
+                "Key length must be in 128-bit."
```
Comment (yegor256) [inline src/main/java/org/takes/facets/auth/codecs/CcAES.java]: I don't understand this message. It should be either "must be 128 bits length", or "in 128-bit chunks", but in this case the `if` condition is wrong

File: src/main/java/org/takes/facets/auth/codecs/CcAES.java [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+            );
+        }
+        this.origin = codec;
+        final byte[] passcode = new byte[key.length];
+        System.arraycopy(key, 0, passcode, 0, key.length);
+        final SecureRandom random = new SecureRandom();
+        final byte[] ivbytes = new byte[block];
+        random.nextBytes(ivbytes);
+        final AlgorithmParameterSpec spec =
+                new IvParameterSpec(ivbytes);
```
Comment (yegor256) [inline src/main/java/org/takes/facets/auth/codecs/CcAES.java]: why so many indentation spaces here, why not 4?

File: pom.xml [hunk 11]
```diff
@@ -110,6 +110,11 @@
             <optional>true</optional>
         </dependency>
         <dependency>
+            <groupId>com.jcabi</groupId>
+            <artifactId>jcabi-aspects</artifactId>
+            <optional>true</optional>
```
Comment (yegor256) [inline pom.xml]: what is this for? if you need this for tests, use `test` scope

**Comment by yegor256** [comment]
@super132 a few comment from me above

File: src/main/java/org/takes/facets/auth/codecs/CcAES.java (hunk 9 above)
Comment (super132) [inline src/main/java/org/takes/facets/auth/codecs/CcAES.java]: @yegor256, what I mean is "must be 128 bits length". I will modify the message.

File: pom.xml (hunk 11 above)
Comment (super132) [inline pom.xml]: @yegor256, this is for using ```@Imuutable``` annotation suggested by @ggajos, not for testing.

File: src/main/java/org/takes/facets/auth/codecs/CcAES.java (hunk 10 above)
Comment (super132) [inline src/main/java/org/takes/facets/auth/codecs/CcAES.java]: @yegor256, this is a formatting fault from my IDE. I will correct this.

**Comment by super132** [comment]
@yegor256, I have modified as per your comments. Please have a check. Thanks.

File: pom.xml (hunk 11 above)
Comment (yegor256) [inline pom.xml]: no, we shouldn't use it in this project. any extra dependency should be avoided in this project

File: src/main/java/org/takes/facets/auth/codecs/CcAES.java [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+/**
+ * AES codec which supports 128 bits key.
+ *
+ * <p>The class is immutable and thread-safe.
+ *
+ * @author Jason Wong (super132j@yahoo.com)
+ * @version $Id$
+ * @since 0.13.8
+ */
+@Immutable
```
Comment (yegor256) [inline src/main/java/org/takes/facets/auth/codecs/CcAES.java]: remove it, we don't use these annotations in this project

File: src/main/java/org/takes/facets/auth/codecs/CcAES.java [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+     * Ctor.
+     * @param codec Original codec
+     * @param key The encryption key
+     * @exception IOException errors on creating internal components
+     */
+    public CcAES(final Codec codec, final byte[] key) throws IOException {
+        final int block = 16;
+        if (key.length != block) {
+            throw new IllegalArgumentException(
+                "Key must be in 128 bits length."
```
Comment (yegor256) [inline src/main/java/org/takes/facets/auth/codecs/CcAES.java]: I still don't understand this message

**Comment by yegor256** [comment]
@super132 two comments above

File: src/main/java/org/takes/facets/auth/codecs/CcAES.java (hunk 13 above)
Comment (super132) [inline src/main/java/org/takes/facets/auth/codecs/CcAES.java]: @yegor256, what about "Key must be with a length of 128 bits" ?

**Comment by super132** [comment]
@yegor256 I updated the code to remove the dependency and left you a comment on the wording. Thanks.

**Comment by yegor256** [comment]
@rultor merge

**Comment by davvd** [comment]
@elenavolokhova review this ticket please, for compliance with our [quality rules](http://www.teamed.io/qa.html)

**Comment by davvd** [comment]
@rultor deploy

**Comment by elenavolokhova** [comment]
@davvd Quality is good.

**Comment by davvd** [comment]
> @davvd Quality is good.

@elenavolokhova thanks a lot

**Comment by davvd** [comment]
@ggajos just added 10 mins to @elenavolokhova (for QA), payment ID is `56380382`... Thanks for your contribution, **19 mins** was added to your account, payment ID is `AP-4WG017259E3456330`, 153 hours and 20 mins spent total... you're getting extra minutes here (c=4)... +19 added to your rating, current score is: [+3319](http://www.netbout.com/b/35094?open=rating)

**Pull request merged**